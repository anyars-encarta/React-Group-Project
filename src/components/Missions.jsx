import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchMissions, joinMission, leaveMission } from '../redux/missions/missionsSlice';
import '../styles/missions.css';

const Missions = () => {
  const missions = useSelector((state) => state.missions.missions);
  const dispatch = useDispatch();

  const [loading, setLoading] = useState(true);

  useEffect(() => {
    dispatch(fetchMissions())
      .then(() => {
        setLoading(false);
      })
      .catch((error) => {
        throw new Error('Error fetching missions:', error);
      });
  }, [dispatch]);

  return (
    <div className="missions">
      {loading ? (
        <p className="loading">Loading missions...</p>
      ) : (
        <table className="missions-table">
          <thead>
            <tr>
              <th>Mission</th>
              <th>Description</th>
              <th>Status</th>
              <th> </th>
            </tr>
          </thead>
          <tbody>
            {missions.map((mission) => (
              <tr key={mission.mission_id} className={mission.reserved ? '' : 'joined'}>
                <td className="name">{mission.mission_name}</td>
                <td className="description">{mission.description}</td>
                <td className="status">
                  <p className={mission.reserved ? 'member' : 'no-member'}>
                    {mission.reserved ? 'Active Member' : 'NOT A MEMBER'}
                  </p>
                </td>
                <td className="join">
                  {mission.reserved ? (
                    <button type="button" className="leave-m" onClick={() => dispatch(leaveMission(mission.mission_id))}>
                      Leave Mission
                    </button>
                  ) : (
                    <button type="button" className="join-m" onClick={() => dispatch(joinMission(mission.mission_id))}>
                      Join Mission
                    </button>
                  )}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  );
};

export default Missions;
