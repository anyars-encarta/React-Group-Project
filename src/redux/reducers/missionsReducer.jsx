const initialState = {
  missions: [],
  error: null,
};

const missionsReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'FETCH_MISSIONS_SUCCESS':
      return {
        ...state,
        missions: action.payload,
        error: null,
      };
    case 'FETCH_MISSIONS_FAILURE':
      return {
        ...state,
        missions: [],
        error: action.payload,
      };
    default:
      return state;
  }
};

export default missionsReducer;
