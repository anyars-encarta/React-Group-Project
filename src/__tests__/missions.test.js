import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import Missions from '../components/Missions';
import missionsReducer from '../redux/missions/missionsSlice';

const missionStore = configureStore({
  reducer: {
    missions: missionsReducer,
  },
});

const data = [
  {
    mission_id: 'mission1',
    mission_name: 'Mission Apolo',
    description: 'Mission Mission, Mission Mission Space',
  },
  {
    mission_id: 'mission2',
    mission_name: 'Mission Apolo 2',
    description: 'Mission Mission, Mission Mission Space 2',
  },
  {
    mission_id: 'mission3',
    mission_name: 'Mission Apolo 3',
    description: 'Mission Mission, Mission Mission Space 3',
  },
];

describe('Missions component', () => {
  it('renders a list of missions', async () => {
    global.fetch = jest.fn(() => Promise.resolve({ json: () => Promise.resolve(data) }));
    const { asFragment } = render(
      <Provider store={missionStore}>
        <Missions />
      </Provider>,
    );

    await waitFor(() => expect(screen.getByText('Mission Apolo')).toBeInTheDocument());
    await waitFor(() => expect(screen.getByText('Mission Apolo 2')).toBeInTheDocument());
    await waitFor(() => expect(screen.getByText('Mission Apolo 3')).toBeInTheDocument());

    expect(asFragment()).toMatchSnapshot();
  });
});
