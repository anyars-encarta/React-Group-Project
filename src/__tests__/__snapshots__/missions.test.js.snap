// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Missions component renders a list of missions 1`] = `
<DocumentFragment>
  <div
    class="missions"
  >
    <table
      class="missions-table"
    >
      <thead>
        <tr>
          <th>
            Mission
          </th>
          <th>
            Description
          </th>
          <th>
            Status
          </th>
          <th>
             
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          class="joined"
        >
          <td
            class="name"
          >
            Mission Apolo
          </td>
          <td
            class="description"
          >
            Mission Mission, Mission Mission Space
          </td>
          <td
            class="status"
          >
            <p
              class="no-member"
            >
              NOT A MEMBER
            </p>
          </td>
          <td
            class="join"
          >
            <button
              class="join-m"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr
          class="joined"
        >
          <td
            class="name"
          >
            Mission Apolo 2
          </td>
          <td
            class="description"
          >
            Mission Mission, Mission Mission Space 2
          </td>
          <td
            class="status"
          >
            <p
              class="no-member"
            >
              NOT A MEMBER
            </p>
          </td>
          <td
            class="join"
          >
            <button
              class="join-m"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
        <tr
          class="joined"
        >
          <td
            class="name"
          >
            Mission Apolo 3
          </td>
          <td
            class="description"
          >
            Mission Mission, Mission Mission Space 3
          </td>
          <td
            class="status"
          >
            <p
              class="no-member"
            >
              NOT A MEMBER
            </p>
          </td>
          <td
            class="join"
          >
            <button
              class="join-m"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
